corehq/apps/domain/decorators.py

```diff
@@ -161,22 +166,16 @@ def _inner(request, domain, *args, **kwargs):
                         return HttpResponseForbidden()
 
                 return _inner(request, domain, *args, **kwargs)
-            else:
+            elif allow_sessions:
                 return login_and_domain_required(fn)(request, domain, *args, **kwargs)
+            else:
+                return HttpResponseForbidden()
```

Comment:
I think this else clause will never be executed. It also seems easier to understand if the first condition is inverted (stated positively). Here's a series of transformations that I believe are logically consistent and end up clearer:

```py
if not request.user.is_authenticated() or not allow_sessions:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
elif allow_sessions:
    return login_and_domain_required(fn)(request, domain, *args, **kwargs)
else:
    return HttpResponseForbidden()
```
is the same as (`elif` -> `else` with nested `if`)
```py
if not request.user.is_authenticated() or not allow_sessions:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
else:
    if allow_sessions:
        return login_and_domain_required(fn)(request, domain, *args, **kwargs)
    else:
        return HttpResponseForbidden()
```
is the same as (invert first if condition)
```py
if request.user.is_authenticated() and allow_sessions:
    if allow_sessions:
        return login_and_domain_required(fn)(request, domain, *args, **kwargs)
    else:
        return HttpResponseForbidden()
else:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
```
is the same as (drop redundant condition)
```py
if request.user.is_authenticated() and allow_sessions:
    return login_and_domain_required(fn)(request, domain, *args, **kwargs)
else:
    ... # make _innter that checks stuff, maybe calls fn
    return _inner(request, domain, *args, **kwargs)
```